backends/dpdk/dpdkProgram.cpp

```diff
@@ -36,61 +36,22 @@ const IR::DpdkAsmStatement *ConvertToDpdkProgram::createListStatement(
 }
 
 const IR::DpdkAsmProgram *ConvertToDpdkProgram::create(IR::P4Program *prog) {
-    IR::IndexedVector<IR::DpdkHeaderType> headerType;
-    for (auto kv : structure.header_types) {
-        LOG3("add header type " << kv.second);
-        auto h = kv.second;
-        auto ht = new IR::DpdkHeaderType(h->srcInfo, h->name, h->annotations,
-                                         h->fields);
-        headerType.push_back(ht);
-    }
-    IR::IndexedVector<IR::DpdkStructType> structType;
+    IR::Type_Struct *metadataStruct = nullptr;
+
     for (auto obj : prog->objects) {
         if (auto s = obj->to<IR::Type_Struct>()) {
             if (s->name.name == info->local_metadata_type) {
-                auto *annotations = new IR::Annotations(
-                    {new IR::Annotation(IR::ID("__metadata__"), {})});
-                for (auto anno : s->annotations->annotations)
-                    annotations->add(anno);
-                auto st = new IR::DpdkStructType(s->srcInfo, s->name,
-                                                 annotations, s->fields);
-                structType.push_back(st);
-            } else if (args_struct_map->find(s->name.name) !=
-                       args_struct_map->end()) {
-                auto st = new IR::DpdkStructType(s->srcInfo, s->name,
-                                                 s->annotations, s->fields);
-                structType.push_back(st);
-            } else if (s->name.name == info->header_type) {
-                auto *annotations = new IR::Annotations(
-                    {new IR::Annotation(IR::ID("__packet_data__"), {})});
-                for (auto anno : s->annotations->annotations)
-                    annotations->add(anno);
-                auto st = new IR::DpdkStructType(s->srcInfo, s->name,
-                                                 s->annotations, s->fields);
-                structType.push_back(st);
+                metadataStruct = s->clone();
```

Comment:
No break?

Reply:
This seems a bit brittle; in other backends we use the type of the suitable parameters (e.g., third parameter of ingress) to find the metadata type. But perhaps this will work, I don't remember how local_metadata_type is computed.

Reply:
Yes. Break is needed here. I will update the code.

I used clone here as the metadata struct returned here as program object is const and I need to insert new fields to it. Any other way to do that?

local_metadata_type is collected from the Ingress Parser block parameters in an earlier pass "CollectMetadataHeaderInfo". The information collected in this pass is extensively used in many other passes.

 ```
auto local_metadata = parser->getApplyParameters()->getParameter(2);
 local_metadata_type =
                local_metadata->type->to<IR::Type_Name>()->path->name;
```